### PR TITLE
feat(operator): add support for pod template

### PR DESCRIPTION
**Pull request by jsenko**

### CODE REVIEW

**Comment by jsenko** [comment]
@andreaTP Please review, I'll take a look at your PR soon.

File: operator/controller/src/test/resources/podtemplatespec-test-cases.yaml [hunk 1]
```diff
@@ -0,0 +1,481 @@
...
+                memory: 2Gi
+    expected:
+      metadata: { }
+      spec:
+        containers:
+          - name: test
+            image: test:tag
+            resources:
+              requests:
+                cpu: 750m
```
Comment (andreaTP) [inline operator/controller/src/test/resources/podtemplatespec-test-cases.yaml:477]: Looking at this example, I start to think that the approach used for the probes is the most straightforward to understand and reason about:
if you define a resource request or limit, then it will override the entire object.

We also need to define the semantics for "removing" things.

File: operator/controller/src/main/java/io/apicurio/registry/operator/utils/PodTemplateSpecUtils.java [hunk 2]
```diff
@@ -0,0 +1,147 @@
...
+
+import java.util.ArrayList;
+import java.util.HashMap;
+
+import static io.apicurio.registry.operator.utils.Mapper.duplicate;
+import static io.apicurio.registry.operator.utils.Mapper.toYAML;
+import static io.apicurio.registry.operator.utils.Utils.*;
+import static java.util.Objects.requireNonNull;
+
+public class PodTemplateSpecUtils {
```
[the remark this answers is not in the document]
Comment (jsenko) [reply operator/controller/src/main/java/io/apicurio/registry/operator/utils/PodTemplateSpecUtils.java:17]: `Resource` has a specific meaning in JOSDK. I don't see a problem with `utils`, but I can rename it to `PodTemplateSpecFeature`.

File: operator/controller/src/main/java/io/apicurio/registry/operator/utils/PodTemplateSpecUtils.java [hunk 3]
```diff
@@ -0,0 +1,147 @@
...
+        if (merged.getMetadata() == null) {
+            merged.setMetadata(new ObjectMeta());
+        }
+
+        // .metadata.labels
+        if (merged.getMetadata().getLabels() == null) {
+            merged.getMetadata().setLabels(new HashMap<>());
+        }
+        if (original.getMetadata() != null) {
+            mergeOverride(merged.getMetadata().getLabels(), original.getMetadata().getLabels());
```
[the remark this answers is not in the document]
Comment (jsenko) [reply operator/controller/src/main/java/io/apicurio/registry/operator/utils/PodTemplateSpecUtils.java:49]: I disagree, since it avoids repeating a null check, but I'll inline the implementation if you want.

File: operator/controller/src/test/java/io/apicurio/registry/operator/unit/MergePTSTest.java [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.assertj.core.api.Assertions.fail;
+
+public class MergePTSTest {
+    private static final Logger log = LoggerFactory.getLogger(MergePTSTest.class);
+
+    private static final String TEST_CONTAINER_NAME = "test";
+
+    @Test
+    void shouldReturnAnErrorStatus() {
```
[the remark this answers is not in the document]
Comment (jsenko) [reply operator/controller/src/test/java/io/apicurio/registry/operator/unit/MergePTSTest.java:24]: This is my mistake from copying the test class, I'll rename it.

[the remark this answers is not in the document]
Comment (jsenko) [reply operator/controller/src/test/java/io/apicurio/registry/operator/unit/MergePTSTest.java:24]: I can split the files.

File: operator/controller/src/test/resources/podtemplatespec-test-cases.yaml [hunk 5]
```diff
@@ -0,0 +1,481 @@
+cases:
+
+  - id: 'Empty 1'
```
[the remark this answers is not in the document]
Comment (jsenko) [reply operator/controller/src/test/resources/podtemplatespec-test-cases.yaml:3]: I left them out because as currently implemented, these are taken directly from the spec and do not need to be merged from the factory, and any additional changes that the operator does will be done on top of that.

File: operator/controller/src/test/resources/podtemplatespec-test-cases.yaml (hunk 1 above)
Comment (jsenko) [reply operator/controller/src/test/resources/podtemplatespec-test-cases.yaml:477]: The difference between probes and R&R is that for probes, (I think) you usually define a single kind of probe, and rarely change only some specific setting such as `initialDelaySeconds`. It is also harder to merge than R&L, since they are simple maps.

> We also need to define the semantics for "removing" things.
Can you clarify please?

File: operator/controller/src/test/java/io/apicurio/registry/operator/unit/MergePTSTest.java [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+    public static class TestCases {
+        private List<TestCase> cases;
+    }
+
+    @Getter
+    @Setter
+    public static class TestCase {
+        private String id;
+        private PodTemplateSpec original;
+        private PodTemplateSpec spec;
```
[the remark this answers is not in the document]
Comment (jsenko) [reply operator/controller/src/test/java/io/apicurio/registry/operator/unit/MergePTSTest.java:65]: I think ` spec` makes a lot of sense, since it comes from the spec. `expected` is used on the line below.

File: operator/controller/src/main/resources/k8s/default/app.deployment.yaml [hunk 7]
```diff
@@ -9,6 +9,8 @@ spec:
   template:
     metadata: { }
     spec:
+      # WARNING: Make sure to update the io.apicurio.registry.operator.utils.PodTemplateSpecUtils.merge method
```
[the remark this answers is not in the document]
Comment (jsenko) [reply operator/controller/src/main/resources/k8s/default/app.deployment.yaml:12]: The only think I can think of is to remove the factory loading from file, and manually apply those defaults in code on top of PTS. Is that what you mean?

File: operator/controller/src/main/java/io/apicurio/registry/operator/utils/PodTemplateSpecUtils.java [hunk 8]
```diff
@@ -0,0 +1,147 @@
...
+        // .metadata.annotations
+        if (merged.getMetadata().getAnnotations() == null) {
+            merged.getMetadata().setAnnotations(new HashMap<>());
+        }
+        if (original.getMetadata() != null) {
+            mergeOverride(merged.getMetadata().getAnnotations(), original.getMetadata().getAnnotations());
+        }
+
+        // .spec.containers[name = containerName]
+        var oc = getContainer(original, containerName);
```
[the remark this answers is not in the document]
Comment (jsenko) [reply operator/controller/src/main/java/io/apicurio/registry/operator/utils/PodTemplateSpecUtils.java:61]: It's a second implementation AFAIK, but I'll improve it.

File: operator/controller/src/main/resources/k8s/default/app.deployment.yaml (hunk 7 above)
[the remark this answers is not in the document]
Comment (andreaTP) [reply operator/controller/src/main/resources/k8s/default/app.deployment.yaml:12]: That's one possible approach, yes.

File: operator/controller/src/test/java/io/apicurio/registry/operator/it/PTSITTest.java [hunk 9]
```diff
@@ -0,0 +1,67 @@
...
+                    .withName(registry.getMetadata().getName() + "-ui-service").get().getSpec()
+                    .getClusterIP()).isNotBlank();
+            return true;
+        });
+
+        int appServicePort = portForwardManager
+                .startPortForward(registry.getMetadata().getName() + "-app-service", 8080);
+
+        await().ignoreExceptions().until(() -> {
+            given().get(new URI("http://localhost:" + appServicePort
```
[the remark this answers is not in the document]
Comment (andreaTP) [reply operator/controller/src/test/java/io/apicurio/registry/operator/it/PTSITTest.java:62]: Yup :+1:

File: operator/controller/src/test/java/io/apicurio/registry/operator/unit/MergePTSTest.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (andreaTP) [reply operator/controller/src/test/java/io/apicurio/registry/operator/unit/MergePTSTest.java:24]: That's in the direction of keeping things simple:

- one test that does `assertEquals` that checks the happy path
- one test that does `assertThrows` that check the negative

without mixing the logic of the two code path in one single implementation.
Am I missing something?

File: operator/controller/src/test/java/io/apicurio/registry/operator/unit/MergePTSTest.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (andreaTP) [reply operator/controller/src/test/java/io/apicurio/registry/operator/unit/MergePTSTest.java:65]: Oh, I see why you named it `spec` now ... too many `spec`s around here, PodTemplate`Spec`, the CR `spec` etc. etc.
Naming is hard and easily confusing here, keep it as a nitpick if we don't easily converge.

File: operator/controller/src/test/resources/podtemplatespec-test-cases.yaml (hunk 5 above)
[the remark this answers is not in the document]
Comment (andreaTP) [reply operator/controller/src/test/resources/podtemplatespec-test-cases.yaml:3]: as far as we remember to implement the relevant checks when implementing new features, I'm ok with this logic.

File: operator/controller/src/test/resources/podtemplatespec-test-cases.yaml (hunk 1 above)
Comment (andreaTP) [reply operator/controller/src/test/resources/podtemplatespec-test-cases.yaml:477]: > Can you clarify please?

the practical use case is:
how do I remove the `cpu` limit using `PodTemplateSpec`?

I think I like the most the full override approach used for probes, do you see any issue with that?

Comment (jsenko) [reply operator/controller/src/test/resources/podtemplatespec-test-cases.yaml:477]: In the current approach, users can "remove" the cpu limit by setting a different one, I assume it's very rare where users would want to only remove the limit, but I don't have a problem with that.

File: operator/controller/src/main/java/io/apicurio/registry/operator/utils/PodTemplateSpecUtils.java [hunk 10]
```diff
@@ -0,0 +1,147 @@
...
+        }
+
+        var mc = getContainer(merged, containerName);
+        if (mc == null) {
+            merged.getSpec().getContainers().add(oc);
+        } else {
+
+            // .spec.containers[name = containerName].env
+            if (!isEmpty(mc.getEnv())) {
+                throw new OperatorException("""
```
[the remark this answers is not in the document]
Comment (jsenko) [reply operator/controller/src/main/java/io/apicurio/registry/operator/utils/PodTemplateSpecUtils.java:84]: We would have two places for specifying env. variables with two different behaviors:
- env from the PTS can be overridden by the operator
- env. from the spec.(app/ui).env override values set by the operator.
I think allowing the first case would confuse and might surprise users.

[the remark this answers is not in the document]
Comment (jsenko) [reply operator/controller/src/main/java/io/apicurio/registry/operator/utils/PodTemplateSpecUtils.java:84]: The limitation on image can be removed

File: operator/controller/src/main/java/io/apicurio/registry/operator/utils/PodTemplateSpecUtils.java [hunk 11]
```diff
@@ -0,0 +1,147 @@
...
+            if (!isEmpty(mc.getImage())) {
+                // TODO: Do we want to allow this?
+                throw new OperatorException("""
+                        Field spec.(app/ui).podTemplateSpec.spec.containers[name = %s].image must be empty."""
+                        .formatted(containerName));
+            }
+            mc.setImage(oc.getImage());
+
+            // .spec.containers[name = containerName].ports
+            mergeNotOverride(mc.getPorts(), oc.getPorts(), ContainerPort::getName);
```
[the remark this answers is not in the document]
Comment (jsenko) [reply operator/controller/src/main/java/io/apicurio/registry/operator/utils/PodTemplateSpecUtils.java]: This method is complex enough to keep

[the remark this answers is not in the document]
Comment (andreaTP) [reply operator/controller/src/main/java/io/apicurio/registry/operator/utils/PodTemplateSpecUtils.java]: it's a one liner:

```java
source.forEach(target::putIfAbsent);
```

@jsenko please, let me "Resolve" the conversations, otherwise is really hard for me to validate the progress :pray:

[the remark this answers is not in the document]
Comment (andreaTP) [reply operator/controller/src/main/java/io/apicurio/registry/operator/utils/PodTemplateSpecUtils.java]: oh, this is the overload :sweat_smile: let me have a closer look.

[the remark this answers is not in the document]
Comment (andreaTP) [reply operator/controller/src/main/java/io/apicurio/registry/operator/utils/PodTemplateSpecUtils.java]: Ok, the implementation is something like (not tested):

```java
       source.forEach((sKey, sValue) -> {
            target.forEach((tKey, tValue) -> {
                if (sValue.getName().equals(tValue.getName())) {
                    target.put(sKey, sValue);
                };
            });
        });
```

Doing it I realized that I'm not sure about the semantics, here we are using the name of the port, I think that we should use the key of the map and document the behavior accordingly.
The rationale is:
looking inside the `ContainerPort` object one can argue why we are using the name and not the port number, and we are not considering the other fields(like protocol etc.).
The outern key, instead, is a direct identifier that cannot be duplicated due to yaml not accepting it.

Bonus point:
the current implementation is computationally complex as it does full target table scans, one each source Map.
I don't expect a lot of elements in those maps, so this just a nice side effect.

**Comment by jsenko** [comment]
@andreaTP I've changed the approach to implementing the PTS support. Let me know what you think.

**Comment by jsenko** [comment]
I can close and reopen the PR if you think this is too cluttered.

File: operator/controller/src/test/resources/podtemplatespec-test-cases.yaml (hunk 1 above)
Comment (andreaTP) [reply operator/controller/src/test/resources/podtemplatespec-test-cases.yaml]: > In the current approach, users can "remove" the cpu limit by setting a different one, I assume it's very rare where users would want to only remove the limit

Setting a different value doesn't remove the limit, I have seen a lot of use cases where people are removing the CPU limit to avoid the applications to be throttled(which is a condition hard to identify in a live cluster).

File: operator/controller/src/main/java/io/apicurio/registry/operator/resource/ResourceFactory.java [hunk 12]
```diff
@@ -28,16 +38,157 @@ public class ResourceFactory {
     public static final String UI_CONTAINER_NAME = "apicurio-registry-ui";
 
     public Deployment getDefaultAppDeployment(ApicurioRegistry3 primary) {
-        var r = getDefaultResource(primary, Deployment.class, RESOURCE_TYPE_DEPLOYMENT, COMPONENT_APP);
-        addDefaultLabels(r.getSpec().getTemplate().getMetadata().getLabels(), primary, COMPONENT_APP);
+        var r = new Deployment();
+
+        // Metadata
+        if (r.getMetadata() == null) {
```
Comment (andreaTP) [inline operator/controller/src/main/java/io/apicurio/registry/operator/resource/ResourceFactory.java:44]: are those null checks necessary?
we are building a brand new `Deployment` object every time, they are going to be always `null`

File: operator/controller/src/main/java/io/apicurio/registry/operator/resource/ResourceFactory.java [hunk 13]
```diff
@@ -28,16 +38,157 @@ public class ResourceFactory {
...
+        if (c.getEnv() != null && !c.getEnv().isEmpty()) {
+            throw new OperatorException("""
+                    Field spec.(app/ui).podTemplateSpec.spec.containers[name = %s].env must be empty.  \
+                    Use spec.(app/ui).env to configure environment variables."""
+                    .formatted(APP_CONTAINER_NAME));
+        }
+        if (c.getPorts() == null) {
+            c.setPorts(new ArrayList<>());
+        }
+        mergeNotOverride(c.getPorts(), List.of(new ContainerPortBuilder().withName("http").withProtocol("TCP")
```
Comment (andreaTP) [inline operator/controller/src/main/java/io/apicurio/registry/operator/resource/ResourceFactory.java:93]: no need to merge anything here as we are building a brand new thing.

File: operator/controller/src/main/java/io/apicurio/registry/operator/resource/ResourceFactory.java (hunk 12 above)
Comment (andreaTP) [reply operator/controller/src/main/java/io/apicurio/registry/operator/resource/ResourceFactory.java:44]: the comment applies to all the occurrences.

File: operator/controller/src/main/java/io/apicurio/registry/operator/resource/ResourceFactory.java (hunk 13 above)
Comment (andreaTP) [reply operator/controller/src/main/java/io/apicurio/registry/operator/resource/ResourceFactory.java:93]: the comment applies to all the occurences.

File: operator/controller/src/main/java/io/apicurio/registry/operator/resource/app/AppDeploymentResource.java [hunk 14]
```diff
@@ -99,15 +100,40 @@ public static void addEnvVar(Map<String, EnvVar> map, EnvVar envVar) {
...
         throw new OperatorException(
                 "Container %s not found in Deployment %s".formatted(name, ResourceID.fromResource(d)));
     }
+
+    /**
+     * Get container with a given name from the given PTS.
+     *
+     * @return null when container was not found
+     */
+    public static Container getContainer(PodTemplateSpec pts, String name) {
```
Comment (andreaTP) [inline operator/controller/src/main/java/io/apicurio/registry/operator/resource/app/AppDeploymentResource.java:127]: rename to `getUserDefinedContainer` ?
I understand why you need the two different implementations and that's fine, but the overload force the reader to always check the types.

File: operator/controller/src/main/java/io/apicurio/registry/operator/utils/Utils.java [hunk 15]
```diff
@@ -8,4 +14,42 @@ private Utils() {
     public static boolean isBlank(String value) {
         return value == null || value.isBlank();
     }
+
+    /**
+     * Merge source map into target map, NOT overriding any entries with the same key.
+     *
+     * @param target must not be null
+     */
+    public static <V> void mergeNotOverride(Map<String, V> target, Map<String, V> source) {
```
Comment (andreaTP) [inline operator/controller/src/main/java/io/apicurio/registry/operator/utils/Utils.java:23]: As mentioned, I'm not in favor of introducing those levels of indirection, they are making the code harder to read with very limited in terms of saved locs

File: operator/controller/src/main/java/io/apicurio/registry/operator/utils/Utils.java [hunk 16]
```diff
@@ -8,4 +14,42 @@ private Utils() {
...
+        }
+    }
+
+    /**
+     * Merge source list into target list, NOT overriding any items based on the equality of the extracted
+     * key.
+     *
+     * @param target must not be null
+     */
+    public static <T, K> void mergeNotOverride(List<T> target, List<T> source, Function<T, K> extractKey) {
```
Comment (andreaTP) [inline operator/controller/src/main/java/io/apicurio/registry/operator/utils/Utils.java:36]: I'm convinced this can be done in a simpler way, let me know if I should take a stab at it :+1:

File: operator/controller/src/test/java/io/apicurio/registry/operator/it/PTSITTest.java [hunk 17]
```diff
@@ -0,0 +1,66 @@
...
+import org.slf4j.LoggerFactory;
+
+import java.net.URI;
+
+import static io.restassured.RestAssured.given;
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.awaitility.Awaitility.await;
+
+@QuarkusTest
+public class PTSITTest extends ITBase {
```
Comment (andreaTP) [inline operator/controller/src/test/java/io/apicurio/registry/operator/it/PTSITTest.java:18]: please, let's use human-readable names :pray:

File: operator/controller/src/test/java/io/apicurio/registry/operator/unit/PodTemplateSpecTest.java [hunk 18]
```diff
@@ -0,0 +1,93 @@
...
+import io.fabric8.kubernetes.api.model.ObjectMeta;
+import org.junit.jupiter.params.ParameterizedTest;
+import org.junit.jupiter.params.provider.ArgumentsSource;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.assertj.core.api.Assertions.assertThatThrownBy;
+
+public class PodTemplateSpecTest {
```
Comment (andreaTP) [inline operator/controller/src/test/java/io/apicurio/registry/operator/unit/PodTemplateSpecTest.java:19]: Nice job here! Clear and well organized!

**Review by andreaTP — CHANGES_REQUESTED** [review]
@jsenko I overall like where this is going, let's try to continue here if you are not too bothered by the comments.

I think we can improve by separating the "merge things" logic from the `getDefaultXXXDeployment`.

i.e.:
- `getDefaultXXDeployment` -> constructs in code the default base without any check, from scratch
- `mergeXYZ` -> contains all the logic for merging the default base with the user-defined values

The distinction will also make clear the intentions and semantic decisions over the merging strategies.

Please, remember that, before merging we should have some sort of documentation about the decisions we are taking when merging stuff.

File: operator/controller/src/main/java/io/apicurio/registry/operator/resource/ResourceFactory.java (hunk 13 above)
Comment (jsenko) [reply operator/controller/src/main/java/io/apicurio/registry/operator/resource/ResourceFactory.java:93]: I don't follow. We are building on top of PTs from the spec. The only part we're building from scratch is the Deployment that contains the PTS.

File: operator/controller/src/main/java/io/apicurio/registry/operator/resource/app/AppDeploymentResource.java (hunk 14 above)
Comment (jsenko) [reply operator/controller/src/main/java/io/apicurio/registry/operator/resource/app/AppDeploymentResource.java:127]: I can rename that, but this is really nitpicking. The methods do the same thing but take different first argument, which should be easily distinguishable. And there is also an IDE.

File: operator/controller/src/main/java/io/apicurio/registry/operator/resource/ResourceFactory.java (hunk 12 above)
Comment (jsenko) [reply operator/controller/src/main/java/io/apicurio/registry/operator/resource/ResourceFactory.java:44]: Fai point, but it applies only for occurrences outside the PTS.

File: operator/controller/src/main/java/io/apicurio/registry/operator/resource/ResourceFactory.java (hunk 13 above)
Comment (jsenko) [reply operator/controller/src/main/java/io/apicurio/registry/operator/resource/ResourceFactory.java:93]: We're merging the ports for example.

File: operator/controller/src/main/java/io/apicurio/registry/operator/resource/app/AppDeploymentResource.java (hunk 14 above)
Comment (andreaTP) [reply operator/controller/src/main/java/io/apicurio/registry/operator/resource/app/AppDeploymentResource.java:127]: > but this is really nitpicking

Let me elaborate on this, 2 methods with different semantics named the same is arguably confusing as you need to check the types with the IDE, while, just changing the name, you can review in GH with simple syntax highlighting.

File: operator/controller/src/main/java/io/apicurio/registry/operator/utils/Utils.java (hunk 16 above)
Comment (jsenko) [reply operator/controller/src/main/java/io/apicurio/registry/operator/utils/Utils.java:36]: Can you create a followup PR for this please? PR does not have to be perfect, is should be correct. I'll be happy to review your PR.

File: operator/controller/src/main/java/io/apicurio/registry/operator/utils/Utils.java (hunk 15 above)
Comment (jsenko) [reply operator/controller/src/main/java/io/apicurio/registry/operator/utils/Utils.java:23]: Can it be done in a followup PR from you?

File: operator/controller/src/test/java/io/apicurio/registry/operator/unit/PodTemplateSpecTest.java (hunk 18 above)
Comment (jsenko) [reply operator/controller/src/test/java/io/apicurio/registry/operator/unit/PodTemplateSpecTest.java:19]: Thank you

File: operator/controller/src/main/java/io/apicurio/registry/operator/utils/Utils.java (hunk 16 above)
Comment (andreaTP) [reply operator/controller/src/main/java/io/apicurio/registry/operator/utils/Utils.java:36]: I'm ok with this, agreed.

File: operator/controller/src/main/java/io/apicurio/registry/operator/resource/ResourceFactory.java (hunk 13 above)
Comment (andreaTP) [reply operator/controller/src/main/java/io/apicurio/registry/operator/resource/ResourceFactory.java:93]: I'll follow up in another PR.

File: operator/controller/src/main/java/io/apicurio/registry/operator/resource/ResourceFactory.java (hunk 12 above)
Comment (andreaTP) [reply operator/controller/src/main/java/io/apicurio/registry/operator/resource/ResourceFactory.java:44]: right, when we split the merging logic from the "default" generation everything should flow nicely.

File: operator/controller/src/test/java/io/apicurio/registry/operator/it/PTSITTest.java (hunk 17 above)
Comment (jsenko) [reply operator/controller/src/test/java/io/apicurio/registry/operator/it/PTSITTest.java]: Done.

File: operator/controller/src/main/java/io/apicurio/registry/operator/resource/app/AppDeploymentResource.java (hunk 14 above)
Comment (jsenko) [reply operator/controller/src/main/java/io/apicurio/registry/operator/resource/app/AppDeploymentResource.java]: Done.

File: operator/controller/src/main/java/io/apicurio/registry/operator/resource/ResourceFactory.java (hunk 12 above)
Comment (jsenko) [reply operator/controller/src/main/java/io/apicurio/registry/operator/resource/ResourceFactory.java]: Done.

**Comment by jsenko** [comment]
> @jsenko I overall like where this is going, let's try to continue here if you are not too bothered by the comments.
> 
> I think we can improve by separating the "merge things" logic from the `getDefaultXXXDeployment`.
> 
> i.e.:
> 
>     * `getDefaultXXDeployment` -> constructs in code the default base without any check, from scratch
> 
>     * `mergeXYZ` -> contains all the logic for merging the default base with the user-defined values
> 
> 
> The distinction will also make clear the intentions and semantic decisions over the merging strategies.
> 
> Please, remember that, before merging we should have some sort of documentation about the decisions we are taking when merging stuff.

Updated.

**Comment by jsenko** [comment]
> Please, remember that, before merging we should have some sort of documentation about the decisions we are taking when merging stuff.

We do not have docs ported yet. what kind of documentation do you have in mind?

**Comment by andreaTP** [comment]
> What kind of documentation do you have in mind?

Even a section in the Readme of the sub-module is fine by me.

> The merge logic is stupidly simple

It's "stupidly simple" to you that have written the code, knows where to look for finding information, knows how to use the IDE etc. etc. I think we can improve in inclusiveness here and allow anyone looking at this PR to access a description in plain English of what we are trying to achieve.

All in all, it can be easy or complicated, but there is no documented decision about the tradeoffs we took:
- merging containers based on names
- merging ports based on the name of the port
- rewriting completely probes when defined in PodTemplate
- etc.

I want to be able to read a spec, reason about it, and verify the implementation.
Writing the code and extracting the rationale after the fact may incur in documented bugs and misinterpretations.

Once we have proper documentation in place we can move the text around.

File: operator/controller/src/main/java/io/apicurio/registry/operator/resource/ResourceFactory.java [hunk 19]
```diff
@@ -28,19 +37,128 @@ public class ResourceFactory {
...
+            c.setPorts(new ArrayList<>());
+        }
+        mergeNotOverride(c.getPorts(), ports, ContainerPort::getName);
+        if (c.getReadinessProbe() == null) {
+            c.setReadinessProbe(readinessProbe);
+        }
+        if (c.getLivenessProbe() == null) {
+            c.setLivenessProbe(livenessProbe);
+        }
+        if (c.getResources() == null) {
```
Comment (andreaTP) [inline operator/controller/src/main/java/io/apicurio/registry/operator/resource/ResourceFactory.java:151]: After a bit of thinking about it, I think the logic here should be:
if `resources` is defined in the `PodTemplateSpec` it fully overrides the defaults.
This allows people to remove the limits when necessary.

File: operator/controller/src/main/java/io/apicurio/registry/operator/utils/Utils.java (hunk 16 above)
Comment (andreaTP) [reply operator/controller/src/main/java/io/apicurio/registry/operator/utils/Utils.java:36]: Follow up PR created: https://github.com/jsenko/apicurio-registry/pull/1055

File: operator/controller/src/main/java/io/apicurio/registry/operator/resource/ResourceFactory.java (hunk 19 above)
Comment (jsenko) [reply operator/controller/src/main/java/io/apicurio/registry/operator/resource/ResourceFactory.java:156]: Done.

**Comment by jsenko** [comment]
@andreaTP I've added information to the README.

File: operator/controller/README.md [hunk 20]
```diff
@@ -0,0 +1,57 @@
...
+- run `eval $(minikube -p minikube docker-env)` in the teerminal yuo are going to build the container image
+- build the image `mvn clean package -Dquarkus.container-image.build=true -f operator/controller/pom.xml`
+- run the tests `mvn verify -f operator/controller/pom.xml -Dtest.operator.deployment=remote -Dquarkus.kubernetes.deployment-target=minikube -DskipOperatorTests=false`
+
+#### Testing of Ingresses
+
+To allow the testing of Ingresses when using minikube, run:
+  ```shell
+  minikube addons enable ingress
+  minikube tunnel
```
Comment (andreaTP) [inline operator/controller/README.md:38]: side note: this mechanism seems to be not enough on my machine, but I have not narrowed it down.

Comment (jsenko) [reply operator/controller/README.md:38]: hmm it works for me. Maybe try with the latest version?

File: operator/controller/README.md [hunk 21]
```diff
@@ -0,0 +1,57 @@
...
+  ```
+
+On other clusters, you might need to provide `test.operator.ingress-host` property that contains the base hostname from  where applications on your cluster are accessible.
+
+If your cluster does not have an accessible ingress host, you can skip them using `test.operator.ingress-skip=true` (**not recommended**).
+
+## PodTemplateSpec merging
```
Comment (andreaTP) [inline operator/controller/README.md:45]: Thanks :+1:

**Review by andreaTP — APPROVED** [review]
LGTM, good to go, thanks for bearing with me @jsenko
